Fix Powershell Progress stream leak to stderr for normal and elevated commands

DanHam opened the pull request:
Powershell leaks any output sent to its Progress stream to stderr when using the `-EncodedCommand` option. Note that the leak to stderr still occurs despite setting `$ProgressPreference` to `SilentlyContinue`.

Observations:

* v0.12.0 did not suffer with this issue for normal commands but did for elevated commands
* By default with v0.12.0 ProgressPreference was Continue for normal commands and SilentlyContinue for elevated commands

I've tested the changes with a Windows 2012r2 build and a Windows 2016 build and this has solved the issue. Clearly more testing with other Windows/Powershell variants would be welcome...

Closes #4322

DanHam commented:
@SwampDragons I've noticed that we are now getting some unwanted output on the CLI immediately after WinRM connects that we didn't get before.

```
==> virtualbox-iso: Waiting for WinRM to become available...
    virtualbox-iso: WinRM connected.
    virtualbox-iso: #< CLIXML
    virtualbox-iso: <Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj><Obj S="progress" RefId="1"><TNRef RefId="0" /><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>
```

I believe this is due to [this commit](https://github.com/hashicorp/packer/commit/959db1ac169166a16b4b5207d329765a8bd4f87a). This makes use of the masterzen/WinRM code [here](https://github.com/hashicorp/packer/blob/4ce381cae720dd8c052ed76b6274dba15b7f8a81/vendor/github.com/masterzen/winrm/powershell.go#L22). As you can see the command has the troublesome -EncodedCommand flag which causes the leak of the Progress stream to StdErr.

SwampDragons commented:
Thanks for reopening!  I'll be sure to take a look at the echo test leak once we get this squared away.

SwampDragons commented:
Worth noting that when I checked this I made sure to use weirdo environment variables that required powershell escaping. Looks like they behave the same in this PR as on master, so there shouldn't be any backwards incompatibility except for where people have their own custom execute_command.

Still, since custom execute_commands will likely be affected, we'll need to save this for 1.2.0

DanHam commented:
@SwampDragons 

> EDIT: I checked again and the environment variables require different escape sequences.

OK. So I am assuming this is because of the change to enclose env vars with single quotes instead of double quotes?  EnvVarFormat and ElevatedEnvVarFormat?

This change was suggested by taliesins because PS code embedded in environment variables is executed by the provisioner when double quotes are used. e.g.

```
    { "type": "powershell",
      "environment_vars": [
        "FOO=$(Write-Host \"Output from PS Code in EnvVars should not appear on packer console\")"
      ],
      "inline": "write-output \"...But it does because of the double quotes!\""
    },
    ...
```
More generally of course, _any_ string making up env vars, usernames, passwords etc that contain characters special to powershell can cause issues e.g. passwords containing dollar signs. Also of concern are backticks, use of double quotes within double quotes, and use of single quotes within single quotes.

I was taking a look at the discussion going on in #5259 and the use of single quotes -vs- double quotes and the problems it was causing. It's pretty similar to the issues we're having here.

I was actually working on some code that would automatically escape chars special to PS so that users wouldn't have to worry about escaping at all for PS. Again this would be backwards incompatible since users that were escaping special characters now would then not need to. However, I think this is preferable to the situation we have now and would ultimately make working with the PS provisioner much easier.

I think it would be best if we rebase and take out this [commit](https://github.com/hashicorp/packer/pull/5365/commits/884f0ba52f8406c28182bea5864e85834c58caf8). This would mean we aren't introducing any backwards incompatibilities and can get this PR merged. What do you think?

In the meantime I'll try and put together some demo code for the auto-escaping I was talking about...

DanHam commented:
@SwampDragons Following on from the discussion above, I've now updated some demo code for you to look at - see #5376.

SwampDragons commented:
Awesome! This PR is first on my list after I get through my weekend-emails.

DanHam commented:
Hi @SwampDragons. Thanks, that's great... but you might want to check out #5368 and the associated PR #5369 first as there is an issue that should probably be fixed first.

After that I'm happy to try and make the changes discussed above to this PR so we can get this one out the way e.g. 

* Take out the commit that changed the env var formats to single quotes - this will ensure we maintain backwards compatibility for the time being
* Include your recommendations on the ```-executionpolicy``` flag for the normal and elevated commands

I've submitted another PR (#5376) that includes the code I was talking about for auto escaping chars in user env vars, and elevated usernames and password strings so that you (and @mwhooker) can take a look at the proposed changes.

Cheers Dan

SwampDragons commented:
> I think it would be best if we rebase and take out this commit. This would mean we aren't introducing any backwards incompatibilities and can get this PR merged. What do you think?

Yep. Agreed.

SwampDragons commented:
Your escape sequence PR looks good to me and I've marked it for 1.2.0 so we don't forget it. I've merged your fix for the dotfile sequencing.  And I think that means this PR is ready, pending the rebase, quoting revert, and `executionpolicy` addition.

DanHam commented:
> ...I think that means this PR is ready, pending the rebase, quoting revert, and executionpolicy addition.

@SwampDragons Great, thank you! So hopefully all done now. I've rebased, reverted the change to single quotes and added ```-executionpolicy bypass``` to the to the ```execute_command``` and ```elevated_execute_command``` as requested.

SwampDragons commented on provisioner/powershell/provisioner.go:
```diff
@@ -112,19 +113,19 @@ func (p *Provisioner) Prepare(raws ...interface{}) error {
 
 	if p.config.EnvVarFormat == "" {
-		p.config.EnvVarFormat = `$env:%s="%s"; `
+		p.config.EnvVarFormat = `$env:%s=\"%s\"; `
```
did you intend to escape these but not the elevated env vars?

DanHam replied:
Yes - that's right. This is due to the difference in how we now incorporate env vars in normal -vs- elevated PowerShell commands.

For the normal command, the env vars are incorporated directly in the command string. _Since the command string itself makes use of double quotes_, we need to escape the quotes surrounding the env var values so we end up with a command something like:

```
powershell -executionpolicy bypass "& { if (Test-Path ... blah};$env:FOO=\"bar\"; $env:BAR=\"baz\";...;  &'c:/Windows/Temp/script-59ca48a4-f4d6-f919-4653-6b3a95e3bb48.ps1';exit $LastExitCode }"
```

We don't need to escape the double quotes surrounding the env var values for elevated commands. Uploading the env vars into a file and dot sourcing them means the env vars don't appear directly in the elevated command string and so we don't need to escape the double quotes.

Hope I've made sense there...

Richard-Payne commented:
Is this actually done?  It's marked for milestone v1.1.2 but I'm still seeing this on v1.4.2.

DanHam commented:
@Richard-Payne It was definately merged and solved the original issue we were experiencing way back when... 

I just ran an old test template for building a Windows 2016 instance on AWS - see [HERE](https://github.com/DanHam/packer-testing/tree/master/aws/windows).

I'm not seeing any Progress stream leaks from the Powershell scripts run by that template. I do however see a leak when Packer first connects over WinRM:

```
==> amazon-ebs: Waiting for WinRM to become available...
==> amazon-ebs: #< CLIXML
    amazon-ebs: WinRM connected.
==> amazon-ebs: <Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj S="progress" RefId="0"><TN RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj><Obj S="progress" RefId="1"><TNRef RefId="0" /><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first use.</AV><AI>0</AI><Nil /><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD> </SD></PR></MS></Obj></Objs>
==> amazon-ebs: Connected to WinRM!
```

Is this what you are referring to?

If so this is a known issue - see my original comment above on this [HERE](https://github.com/hashicorp/packer/pull/5365#issuecomment-330946425). This is the Progress stream leak from an internal test Packer makes to ensure it is actually connected over WinRM and is able to run commands.

Richard-Payne commented:
Yes, exactly that

On 14 August 2019 17:58:43 BST, Dan <notifications@github.com> wrote:
>@Richard-Payne It was definately merged and solved the original issue
>we were experiencing way back when... 
>
>I just ran an old test template for building a Windows 2016 instance on
>AWS - see
>[HERE](https://github.com/DanHam/packer-testing/tree/master/aws/windows).
>
>I'm not seeing any Progress stream leaks from the Powershell scripts
>run by that template. I do however see a leak when Packer first
>connects over WinRM:
>
>```
>==> amazon-ebs: Waiting for WinRM to become available...
>==> amazon-ebs: #< CLIXML
>    amazon-ebs: WinRM connected.
>==> amazon-ebs: <Objs Version="1.1.0.1"
>xmlns="http://schemas.microsoft.com/powershell/2004/04"><Obj
>S="progress" RefId="0"><TN
>RefId="0"><T>System.Management.Automation.PSCustomObject</T><T>System.Object</T></TN><MS><I64
>N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for first
>use.</AV><AI>0</AI><Nil
>/><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD>
></SD></PR></MS></Obj><Obj S="progress" RefId="1"><TNRef RefId="0"
>/><MS><I64 N="SourceId">1</I64><PR N="Record"><AV>Preparing modules for
>first use.</AV><AI>0</AI><Nil
>/><PI>-1</PI><PC>-1</PC><T>Completed</T><SR>-1</SR><SD>
></SD></PR></MS></Obj></Objs>
>==> amazon-ebs: Connected to WinRM!
>```
>
>Is this what you are referring to?
>
>If so this is a known issue - see my original comment above on this
>[HERE](https://github.com/hashicorp/packer/pull/5365#issuecomment-330946425).
>This is the Progress stream leak from an internal test Packer makes to
>ensure it is actually connected over WinRM and is able to run commands.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/hashicorp/packer/pull/5365#issuecomment-521330135

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

Richard-Payne commented:
@DanHam is there a fix for this?  I've read through the comments in this thread again and I still can't get a handle on the current state of the problem.

DanHam commented:
@Richard-Payne Sorry for the late response.

Unfortunately not. The Progress stream leak that you are seeing is not related to the issues that this PR fixed.

We were aware this 'additional' problem at the time but a fix was never implemented. If this is causing you issues I think the best bet would be to open up a new issue to make the Packer developers aware.

Richard-Payne commented:
ok, no problem.  Thanks.
